File: apps/Ingestion Plane/Quarry-v2/.claude/worktrees/vigilant-banach-6933aa/pkg/quarrycontracts/policy.go
package quarrycontracts

type RunPolicy struct {
	Concurrency Concurrency  `json:"concurrency"`
	Delay       Delay        `json:"delay"`
	Retry       Retry        `json:"retry"`
	Proxy       Proxy        `json:"proxy"`
	Robots      RobotsMode   `json:"robots"`
	Ordering    Ordering     `json:"ordering"`
	Block       BlockPolicy  `json:"block"`
	Checkpoint  Checkpoint   `json:"checkpoint"`
	Determinism Determinism  `json:"determinism"`
}

type Concurrency struct {
	PerRun    uint32 `json:"per_run"`
	PerDomain uint32 `json:"per_domain"`
}

type Delay struct {
	MinMs  uint32 `json:"min_ms"`
	MaxMs  uint32 `json:"max_ms"`
	Jitter bool   `json:"jitter"`
}

type Retry struct {
	Max     uint32      `json:"max"`
	Backoff BackoffKind `json:"backoff"`
	BaseMs  uint32      `json:"base_ms"`
}

type BackoffKind string
const (
	BackoffFixed  BackoffKind = "fixed"
	BackoffLinear BackoffKind = "linear"
	BackoffExp    BackoffKind = "exp"
)

type Proxy struct {
	Strategy  ProxyStrategy `json:"strategy"`
	Pool      *string       `json:"pool,omitempty"`
	StickyKey *string       `json:"sticky_key,omitempty"`
}

type ProxyStrategy string
const (
	ProxyRotate ProxyStrategy = "rotate"
	ProxySticky ProxyStrategy = "sticky"
	ProxyNone   ProxyStrategy = "none"
)

type RobotsMode string
const (
	RobotsStrict  RobotsMode = "strict"
	RobotsRespect RobotsMode = "respect"
	RobotsIgnore  RobotsMode = "ignore"
)

type Ordering string
const (
	OrdFifo     Ordering = "fifo"
	OrdPriority Ordering = "priority"
	OrdLifo     Ordering = "lifo"
)

type BlockPolicy struct {
	On     BlockTrigger `json:"on"`
	Action BlockAction  `json:"action"`
}

type BlockTrigger string
const (
	TrigChallenge   BlockTrigger = "challenge"
	TrigStatus429   BlockTrigger = "status_429"
	TrigStatus403   BlockTrigger = "status_403"
	TrigSuspectBot  BlockTrigger = "suspected_bot"
)

type BlockAction string
const (
	BlkEscalate BlockAction = "escalate"
	BlkRetry    BlockAction = "retry"
	BlkAbort    BlockAction = "abort"
)

type Checkpoint struct {
	EveryNPages *uint32 `json:"every_n_pages,omitempty"`
	EveryS      *uint32 `json:"every_s,omitempty"`
}

type Determinism string
const (
	DetStrict    Determinism = "strict"
	DetBestEff   Determinism = "best_effort"
	DetOff       Determinism = "off"
)

func DefaultRunPolicy() RunPolicy {
	n := uint32(50); s := uint32(60)
	return RunPolicy{
		Concurrency: Concurrency{PerRun: 8, PerDomain: 2},
		Delay:       Delay{MinMs: 500, MaxMs: 2000, Jitter: true},
		Retry:       Retry{Max: 3, Backoff: BackoffExp, BaseMs: 1000},
		Proxy:       Proxy{Strategy: ProxyNone},
		Robots:      RobotsRespect,
		Ordering:    OrdFifo,
		Block:       BlockPolicy{On: TrigChallenge, Action: BlkEscalate},
		Checkpoint:  Checkpoint{EveryNPages: &n, EveryS: &s},
		Determinism: DetBestEff,
	}
}

File: apps/Ingestion Plane/Quarry-v2/.claude/worktrees/vigilant-banach-6933aa/crates/quarry-core/src/policy.rs
//! Run + domain policy. Mirrors CONTRACTS §7.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunPolicy {
    pub concurrency: Concurrency,
    pub delay: Delay,
    pub retry: Retry,
    pub proxy: Proxy,
    pub robots: RobotsMode,
    pub ordering: Ordering,
    pub block: BlockPolicy,
    pub checkpoint: Checkpoint,
    pub determinism: Determinism,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Concurrency {
    pub per_run: u32,
    pub per_domain: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Delay {
    pub min_ms: u32,
    pub max_ms: u32,
    pub jitter: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Retry {
    pub max: u32,
    pub backoff: BackoffKind,
    pub base_ms: u32,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BackoffKind {
    Fixed,
    Linear,
    Exp,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Proxy {
    pub strategy: ProxyStrategy,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pool: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sticky_key: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProxyStrategy {
    Rotate,
    Sticky,
    None,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RobotsMode {
    Strict,
    Respect,
    Ignore,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Ordering {
    Fifo,
    Priority,
    Lifo,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockPolicy {
    pub on: BlockTrigger,
    pub action: BlockAction,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BlockTrigger {
    Challenge,
    Status429,
    Status403,
    SuspectedBot,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BlockAction {
    Escalate,
    Retry,
    Abort,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Checkpoint {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub every_n_pages: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub every_s: Option<u32>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Determinism {
    Strict,
    BestEffort,
    Off,
}

impl Default for RunPolicy {
    fn default() -> Self {
        Self {
            concurrency: Concurrency {
                per_run: 8,
                per_domain: 2,
            },
            delay: Delay {
                min_ms: 500,
                max_ms: 2000,
                jitter: true,
            },
            retry: Retry {
                max: 3,
                backoff: BackoffKind::Exp,
                base_ms: 1000,
            },
            proxy: Proxy {
                strategy: ProxyStrategy::None,
                pool: None,
                sticky_key: None,
            },
            robots: RobotsMode::Respect,
            ordering: Ordering::Fifo,
            block: BlockPolicy {
                on: BlockTrigger::Challenge,
                action: BlockAction::Escalate,
            },
            checkpoint: Checkpoint {
                every_n_pages: Some(50),
                every_s: Some(60),
            },
            determinism: Determinism::BestEffort,
        }
    }
}
